robot_calibration/src/sensor_data_generator.cpp

```diff
@@ -0,0 +1,64 @@
+#include "ros/ros.h"
+#include "sensor_msgs/JointState.h"
+#include <cmath>
+#include "yaml-cpp/yaml.h"
+#include <fstream>
+
+class Listener
+{  
+  public:
+    void calculateAndWriteData(const sensor_msgs::JointState::ConstPtr& msg);
+    std::string timestamp = std::to_string(ros::Time::now().toSec());
+};
+
+
+void Listener::calculateAndWriteData(const sensor_msgs::JointState::ConstPtr& msg)
+{
+  double angles[2];
+  angles[0] = msg->position[0];
+  angles[1] = msg->position[1];
+
+  double link_1_length, link_2_length;
+  link_1_length = 1.8;
+  link_2_length = 1.0;
+
+  double ee_position[] = {(link_1_length * cos(angles[0])) + (link_2_length * cos(angles[0] + angles[1])) , (link_1_length * sin(angles[0])) + (link_2_length * sin(angles[0] + angles[1]))}; 
+  
+  angles[0] += 0.044; //theta1 offset
+  angles[1] += 0.013; //theta2 offset
+
+  YAML::Emitter yaml_out_stream;
+  std::ofstream outfile;
+
+  yaml_out_stream << YAML::BeginMap;
+  yaml_out_stream << YAML::Key  << "Angles";
+  yaml_out_stream << YAML::Value;
+  yaml_out_stream << YAML::Flow;
+  yaml_out_stream << YAML::BeginSeq << angles[0] << angles[1] << YAML::EndSeq;
+  yaml_out_stream << YAML::Key  << "End Effector Position";
+  yaml_out_stream << YAML::Value;
+  yaml_out_stream << YAML::Flow;
+  yaml_out_stream << YAML::BeginSeq <<  ee_position[0] << ee_position[1] << YAML::EndSeq;
+  yaml_out_stream << YAML::EndMap;
+
+  outfile.open("/home/jad/catkin_ws/src/Robot-Calibration-Deep-Dives-DP/robot_calibration/sensor_data/sensor_data" + this->timestamp + ".yaml", std::fstream::app);
```

Comment:
Don't hardcode this path. You can use `ros::package::getPath("robot_calibration")` to find the package path and then create folders/files accordingly. (see [here](http://docs.ros.org/indigo/api/roslib/html/c++/namespaceros_1_1package.html#ae9470dd201aa4e66abb833e710d812a4)) Look into using the C++ [filesystem](https://en.cppreference.com/w/cpp/filesystem/path/relative_path) library to do this as well, since this has functions that check if a directory exists etc.

Reply:
To follow this up, the name of the file which is being saved to should be generated in the class constructor and stored in a member variable, since it avoids having to do these string operations. It might also make it easier to make changes later on if needed.